### PR TITLE
txnengine: shard policy changes

### PR DESCRIPTION
**Pull request by reusee**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:
txnstorage: clean batch after HandleRead
txnengine: rename ShardToSingleStatic to NoShard
txnengine: add NewDefaultShardPolicy, FallbackShard, HashShard

### CODE REVIEW

File: pkg/txn/storage/txn/batch.go [hunk 1]
```diff
@@ -28,8 +28,15 @@ type BatchIter func() (tuple []Nullable)
 func NewBatchIter(b *batch.Batch) BatchIter {
 	i := 0
 	iter := func() (tuple []Nullable) {
-		if i >= b.Vecs[0].Length() {
-			return
+		for {
```
Comment (fengttt) [inline pkg/txn/storage/txn/batch.go:31]: I am not a fan of the Iter idiom.    Caller/Caller inversed and there is no way to vectorwise anything.

File: pkg/vm/engine/txn/shard_hash.go [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+		l := vec.Length() * size
+		data := unsafe.Slice((*byte)(vector.GetPtrAt(vec, 0)), l)
+		return data[i*size : (i+1)*size], nil
+	}
+	return vec.GetBytes(int64(i)), nil
+}
+
+type Nullable struct {
+	IsNull bool
+	Value  any
```
Comment (fengttt) [inline pkg/vm/engine/txn/shard_hash.go:286]: Do you have to use this Nullable?    Why don't just check vector?

That is , work on whole vector, using nulls and MustTCol.

File: pkg/txn/storage/txn/batch.go (hunk 1 above)
Comment (reusee) [reply pkg/txn/storage/txn/batch.go:31]: The mem engine is row-based, so batches need to be iterated by rows to operate.

File: pkg/vm/engine/txn/shard_hash.go (hunk 2 above)
Comment (reusee) [reply pkg/vm/engine/txn/shard_hash.go:286]: The mem engine is row-based and rows need to be updated individually. I don't think vectors are suitable in row-based structures.

Comment (reusee) [reply pkg/vm/engine/txn/shard_hash.go:286]: I know what you mean. I'll change to use nulls to mask elements.

**Review by nnsgmsone — APPROVED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

**Pull request merged**